qualtran/bloqs/chemistry/pbc/first_quantization/prepare_nu.py

```diff
@@ -0,0 +1,288 @@
+    $$
+        \frac{1}{\sqrt{2^{n_p + 2}}} \sum_{\mu=2}^{n_p+1} \sqrt{2^\mu}
+        |0\dots0\underbrace{1\dots1}{\mu}\rangle
+    $$
+
+    Args:
+        num_bits_p: The number of bits to represent each dimension of the momentum register.
+
+    Registers:
+        mu: the register to prepare the superposition over.
+
+    References:
+        [Fault-Tolerant Quantum Simulations of Chemistry in First Quantization](
+            https://arxiv.org/abs/2105.12767) page 21, Eq 77.
+    """
+    num_bits_p: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build(mu=self.num_bits_p)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(Toffoli(), (self.num_bits_p - 1))}
+
+
+@frozen
+class PrepareNuSuperPositionState(Bloq):
+    r"""Prepare the superposition over $\nu$ following the creation of the $|\mu\rangle$ state.
+
+    Prepares the state in Eq. 78
+
+    $$
+        \frac{1}{\sqrt{2^{n_p + 2}}} \sum_{\mu=2}^{n_p+1}
+        \sum_{\nu_{x,y,z}=-(2^{\mu-1}-1)}^{2^{\mu-1}-1}
+        \frac{1}{2\mu}
+        |\mu\rangle|\nu_x\rangle|\nu_y\rangle|\nu_z\rangle
+    $$
+
+    Args:
+        num_bits_p: The number of bits to represent each dimension of the momentum register.
+
+    Registers:
+        mu: the register one-hot encoded $\mu$ register from eq 77.
+        nu: the momentum transfer register.
+
+    References:
+        [Fault-Tolerant Quantum Simulations of Chemistry in First Quantization](
+            https://arxiv.org/abs/2105.12767) page 21, Eq 78.
+    """
+    num_bits_p: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [Register("mu", self.num_bits_p), Register("nu", self.num_bits_p + 1, shape=(3,))]
+        )
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        # controlled hadamards which cannot be inverted at zero Toffoli cost.
+        return {(Toffoli(), (3 * (self.num_bits_p - 1)))}
+
+
+@frozen
+class FlagZeroAsFailure(Bloq):
+    r"""Bloq to flag if minus zero appears in the $\nu$ state.
+
+    Args:
+        num_bits_p: The number of bits to represent each dimension of the momentum register.
+
+    Registers:
+        nu: the momentum transfer register.
+        flag_minus_zero: a flag bit for failure of the state preparation.
+
+    References:
+        [Fault-Tolerant Quantum Simulations of Chemistry in First Quantization](
+            https://arxiv.org/abs/2105.12767) page 21, Eq 80.
+    """
+    num_bits_p: int
+    adjoint: bool = False
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [Register("nu", self.num_bits_p + 1, shape=(3,)), Register("flag_minus_zero", 1)]
+        )
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        if self.adjoint:
+            # This can be inverted with cliffords.
+            return {(Toffoli(), 0)}
```

Comment:
hmm. I think the intent was to just omit zero-count dependencies. Does this actually work? I guess if it's omitted then the bloq itself might be considered a leaf bloq?

Reply:
I think it works. I didn't see anything odd. I will just return an empty set.